pallets/dmp-queue/src/lib.rs

```diff
@@ -0,0 +1,786 @@
+//! execution (by `on_idle` or another `handle_dmp_messages` call). Individual overweight messages
+//! are scheduled into a separate queue that is only serviced by explicit extrinsic calls.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_std::{prelude::*, convert::TryFrom};
+use cumulus_primitives_core::relay_chain::BlockNumber as RelayBlockNumber;
+use cumulus_primitives_core::DmpMessageHandler;
+use codec::{Encode, Decode};
+use sp_runtime::RuntimeDebug;
+use xcm::{VersionedXcm, v0::{Xcm, Junction, Outcome, ExecuteXcm, Error as XcmError}};
+use frame_support::{traits::EnsureOrigin, dispatch::Weight, weights::constants::WEIGHT_PER_MILLIS};
+pub use pallet::*;
+
+#[derive(Copy, Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+pub struct ConfigData {
+	/// The maximum amount of weight any individual message may consume. Messages above this weight
+	/// go into the overweight queue and may only be serviced explicitly by the
+	/// `ExecuteOverweightOrigin`.
+	max_individual: Weight,
+}
+
+impl Default for ConfigData {
+	fn default() -> Self {
+		Self {
+			max_individual: 10 * WEIGHT_PER_MILLIS,	// 10 ms of execution time maximum by default
+		}
+	}
+}
+
+/// Information concerning our message pages.
+#[derive(Copy, Clone, Eq, PartialEq, Default, Encode, Decode, RuntimeDebug)]
+pub struct PageIndexData {
+	/// The lowest used page index.
+	begin_used: PageCounter,
+	/// The lowest unused page index.
+	end_used: PageCounter,
+	/// The number of overweight messages ever recorded (and thus the lowest free index).
+	overweight_count: OverweightIndex,
+}
+
+/// Simple type used to identify messages for the purpose of reporting events. Secure if and only
+/// if the message content is unique.
+pub type MessageId = [u8; 32];
+
+/// Index used to identify overweight messages.
+pub type OverweightIndex = u64;
+
+/// Index used to identify normal pages.
+pub type PageCounter = u32;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	/// The module configuration trait.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type XcmExecutor: ExecuteXcm<Self::Call>;
+
+		/// Origin which is allowed to execute overweight messages.
+		type ExecuteOverweightOrigin: EnsureOrigin<Self::Origin>;
+	}
+
+	/// The configuration.
+	#[pallet::storage]
+	pub(super) type Configuration<T> = StorageValue<_, ConfigData, ValueQuery>;
+
+	/// The page index.
+	#[pallet::storage]
+	pub(super) type PageIndex<T> = StorageValue<_, PageIndexData, ValueQuery>;
+
+	/// The queue pages.
+	#[pallet::storage]
+	pub(super) type Pages<T> = StorageMap<
+		_,
+		Blake2_128Concat,
+		PageCounter,
+		Vec<(RelayBlockNumber, Vec<u8>)>,
```

Comment:
is it possible to already use `BoundedVec` here?

Would prevent having to backport those changes

Reply:
will take a while yet :)